synapse/lib/oauth.py

```diff
@@ -0,0 +1,393 @@
+})
+
+reqValidTokenResponse = s_config.getJsValidator({
+    'type': 'object',
+    'properties': {
+        'access_token': {'type': 'string'},
+        'expires_in': {'type': 'number', 'exclusiveMinimum': 0},
+    },
+    'additionalProperties': True,
+    'required': ['access_token', 'expires_in'],
+})
+
+def normOAuthTokenData(issued_at, data):
+    '''
+    Normalize timestamps to be in epoch millis and set expires_at/refresh_at.
+    '''
+    reqValidTokenResponse(data)
+    expires_in = data['expires_in']
+    return {
+        'access_token': data['access_token'],
+        'expires_in': expires_in,
+        'expires_at': issued_at + expires_in * 1000,
+        'refresh_at': issued_at + (expires_in * REFRESH_WINDOW) * 1000,
+        'refresh_token': data.get('refresh_token'),
+    }
+
+class OAuthMixin(s_nexus.Pusher):
+    '''
+    Mixin for Cells to organize and execute OAuth token refreshes.
+    '''
+
+    async def _initOAuthManager(self):
+
+        slab = self.slab
+        self.oauth_clients = s_lmdbslab.SlabDict(slab, db=slab.initdb('oauth:v2:clients'))     # key=<provider><user>
+        self.oauth_providers = s_lmdbslab.SlabDict(slab, db=slab.initdb('oauth:v2:providers')) # key=<provider>
+
+        self.oauth_sched_map = {}
+        self.oauth_sched_heap = []
+        self.oauth_sched_wake = asyncio.Event()
+        self.onfini(self.oauth_sched_wake.set)
+
+        self.oauth_actviden = self.addActiveCoro(self._runOAuthRefreshLoop)
+
+        # For testing
+        self._oauth_sched_ran = asyncio.Event()
+        self._oauth_sched_empty = asyncio.Event()
+
+    async def _runOAuthRefreshLoop(self):
+        self.oauth_sched_map.clear()
+        self.oauth_sched_heap.clear()
+        self.oauth_sched_wake.clear()
+
+        for provideriden, useriden, clientconf in self.listOAuthClients():
+            self._scheduleOAuthItem(provideriden, useriden, clientconf)
+
+        await self._oauthRefreshLoop()
+
+    def _scheduleOAuthItem(self, provideriden, useriden, clientconf):
+        if not self.isactive:
+            return
+
+        if clientconf.get('error'):
+            return
+
+        if not clientconf.get('refresh_token'):
+            logger.warning(f'OAuth V2 client missing token to schedule refresh provider={provideriden} user={useriden}')
+            return
+
+        refresh_at = clientconf['refresh_at']
+
+        newitem = (refresh_at, provideriden, useriden)
+
+        old_refresh_at = self.oauth_sched_map.get(newitem[1:])
+        if old_refresh_at == refresh_at:
+            return
+
+        if old_refresh_at is not None:
+            # there's an old item for this client in the refresh queue to remove
+            self.oauth_sched_heap.remove((old_refresh_at, *newitem[1:]))
+            heapq.heapify(self.oauth_sched_heap)
+
+        self.oauth_sched_map[newitem[1:]] = refresh_at
+        heapq.heappush(self.oauth_sched_heap, newitem)
+
+        if self.oauth_sched_heap[0] == newitem:
```

Comment:
Can't we do an identify check here since the item in the heap is a tuple who either is or is not the item we made a few lines previously and did not mutate?